RecCalorimeter/src/components/CreateTruthJet.cpp

```diff
@@ -0,0 +1,68 @@
+#include "CreateTruthJet.h"
+
+// std
+#include <vector>
+#include <math.h>
+
+
+
+DECLARE_COMPONENT(CreateTruthJet)
+
+CreateTruthJet::CreateTruthJet(const std::string& name, ISvcLocator* svcLoc) : Transformer(name, svcLoc,
+                    KeyValue("InputCollection", "MCParticles"),
+                    KeyValue("OutputCollection", "TruthJets")) {
+  declareProperty("JetAlg", m_jetAlg, "Name of jet clustering algorithm");
+  declareProperty("JetRadius", m_jetRadius, "Jet clustering radius");
+  declareProperty("MinPt", m_minPt, "Minimum pT for saved jets");
+}
+
+
+
+StatusCode CreateTruthJet::initialize() {
+  clusterer = new ClusterJet(m_jetAlg, m_jetRadius, m_minPt);
+  return clusterer->initialize();
+
+}
+
+colltype_out  CreateTruthJet::operator()(const colltype_in& input) const{
+  std::vector<fastjet::PseudoJet> clustersPJ;
+  int i=0;
+  for(auto particle: input){
+    fastjet::PseudoJet clusterPJ(particle.getMomentum().x, particle.getMomentum().y, particle.getMomentum().z, particle.getEnergy());
+    clusterPJ.set_user_info(new ClusterInfo(i));
+    clustersPJ.push_back(clusterPJ);
+    i++;
+  }
+
+  std::vector<fastjet::PseudoJet> inclusiveJets = clusterer->cluster(clustersPJ);
+  
+
+  edm4hep::ReconstructedParticleCollection edmJets = edm4hep::ReconstructedParticleCollection();
+
+  for(auto cjet : inclusiveJets){
+    edm4hep::MutableReconstructedParticle jet;
+    jet.setMomentum(edm4hep::Vector3f(cjet.px(), cjet.py(), cjet.pz()));
+    jet.setEnergy(cjet.e());
+    jet.setMass(cjet.m());
+
+    std::vector<fastjet::PseudoJet> constits = cjet.constituents();
+    for(auto constit : constits){
+      edm4hep::MutableReconstructedParticle jetInput;
+      jetInput.setMomentum(edm4hep::Vector3f(constit.px(), constit.py(), constit.pz()));
+      jetInput.setEnergy(constit.e());
+      jetInput.setMass(constit.m());
+
+      int index = constit.user_info<ClusterInfo>().index();
+      jetInput.setPDG((input)[index].getPDG());
+
+      jet.addToParticles(jetInput);
```

Comment:
Is'nt this `jetInput` lost in the output file? One should probably store also this new collection of ReconstructedParticle in the output rootfile. Here is an example of `MultiTransformer`: https://github.com/key4hep/k4FWCore/blob/main/test/k4FWCoreTest/src/components/ExampleFunctionalTransformerMultiple.cpp#L51.

I agree that this is not perfect, having a relation between "genJet" and MCParticle would feel more natural but for now there is no such thing AFAIK.

Reply:
Another suggestion (thanks @tmadlener) is to use Association instead of relations (the difference is that associations are separated object with pointers to pairs of objects, used e.g. to link RECO to GEN quantities that should therefore not be used at analysis level). This will 1) avoid loosing information when casting MCParticles to  ReconstructedParticles  2) avoid duplicating content on disk since the associations are pointers. 

One should use `MCRecoParticleAssociation`:
```
edm4hep::MCRecoParticleAssociation:
    Description: "Association between a ReconstructedParticle and the corresponding MCParticle"
    Author: "EDM4hep authors"
    Members:
      - float weight                        // weight of this association
    OneToOneRelations:
     - edm4hep::ReconstructedParticle  rec  // reference to the reconstructed particle
     - edm4hep::MCParticle sim              // reference to the Monte-Carlo particle
``` 

After defining the association object, you can set it with
```
        relation.setSim(MyMCParticle);
        relation.setRec(MyRecoParticle);
```

Reply:
I might be misunderstanding this, but can these associations be used for a one-to-many map, or just a one-to-one relation? It definitely seems better than what I currently have, but it's not obvious to me how this should work in my particular case.

Reply:
Hi Jennifer, very legit question :-) This may change in the future but for now, each entry in the association collection has a single pair. The way to store a "one to many" is therefore to add multiple entries, all having the same `setRec(MyRecoParticle)` but different `setSim(MyMCParticle)`. Storage wise it is ok. CPU wise we loose a bit at analysis stage, where retrieving all MCParticles associated to a GenJet will be done with:

```
for (const auto& assoc : associationColl) {
  if (assoc.getRec() == MyRecoParticle) {
    theMCParticle = rel.getSim();
  }
}
```

Reply:
Also, I guess this will only work for storing particles for truth jets, not for the calo jets? I'd prefer using the same solution for truth and calo jets, since I think this would make things very confusing otherwise.

Reply:
The two solutions are already somewhat different (OneToManyRelations to Cluster vs to Particle) and storing a GEN level information into a RECO object can also be considered very confusing to some extent but I indeed see the 'pragmatic' advantage to be able to retrieve this information based on relation (and not association) in both cases. SO if this is the solution you prefer, let's go for it!

But can you confirm that the objects added with `jet.addToParticles(jetInput);` are accessible from the output rootfile?

Reply:
Yes, I ran a quick test in python where I printed out the energies of the clusters in a jet, and it ran without any problem. For reference, here is the code:

`from podio import root_io

podio_reader = root_io.Reader('output_fullCalo_SimAndDigi.root')

for event in podio_reader.get("events"):
    jets = event.get("Jets")
    for jet in jets:
      clusters = jet.getClusters()
      for cluster in clusters:
        print (cluster.getEnergy())`

So, I think this solution is sufficient for now, though it would be good to eventually have a solution that links the objects instead.

Reply:
Ok for the clusters attached to `CaloJets` but my question was about the `truthJets`: can we access the ReconstructedParticles that created it? By reading the code, it looks like we are missing the saving of a collection containing the `jetInput` in the output rootfile.

Reply:
Sorry for the confusion, and you're right that it was broken for the truth jets. I've switched to using the associations for the truth jets, but not for the reco jets for now.

Reply:
Great! I think the reco jets are fine with the oneToMany association to Clusters. 

But for the truthJets I think we should go for a mutlitransformer since we now have multiple output (not sure what will happen in multi-threaded environment in the current situation). The main change is that `TruthJetParticleAssociation` should appear as a `KeyValues`. Please find an example here: https://github.com/key4hep/k4FWCore/blob/main/test/k4FWCoreTest/src/components/ExampleFunctionalTransformerMultiple.cpp#L51


I tried to run `RecCalorimeter/tests/options/runTruthJetReco.py`, adding

```
################ Output
from Configurables import PodioOutput
out = PodioOutput("out")
out.outputCommands = ["keep *"]
out.filename = "output_truthJets.root"
ApplicationMgr().TopAlg += [out]
```

but the truthJet collection is empty (and the association as well). Could we adapt the test so that it actually produces some truthJets?

Reply:
I've made both of these changes. I haven't added an output root file to the test, but I added it in temporarily to check that it produced truth jets, and that I was able to read the associations.